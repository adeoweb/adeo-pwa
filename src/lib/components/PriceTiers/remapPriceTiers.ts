import { TCurrency } from 'src/lib/constants/currency';
import { TPriceRange, TTierPrice } from 'src/lib/types/graphql/Product';

export type TPriceTier = {
    quantity: number;
    currency: TCurrency;
    value: number;
};

const remapPriceTiers = ({
    priceRange,
    priceTiers = []
}: {
    priceRange?: TPriceRange;
    priceTiers?: TTierPrice[];
}): TPriceTier[] => {
    const items: TPriceTier[] = [];

    if (
        priceRange &&
        priceRange.minimum_price &&
        priceRange.minimum_price.final_price &&
        priceRange.minimum_price.final_price.currency &&
        priceRange.minimum_price.final_price.value
    ) {
        items.push({
            quantity: 1,
            currency: priceRange.minimum_price.final_price.currency,
            value: priceRange.minimum_price.final_price.value
        });
    }

    priceTiers.forEach(item => {
        if (
            item.quantity &&
            item.final_price &&
            item.final_price.currency &&
            item.final_price.value
        ) {
            items.push({
                quantity: item.quantity,
                currency: item.final_price.currency,
                value: item.final_price.value
            });
        }
    });

    return items;
};

export default remapPriceTiers;
