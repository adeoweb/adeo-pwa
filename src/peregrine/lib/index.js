// components
export { default as PeregrineContextProvider } from './PeregrineContextProvider';
