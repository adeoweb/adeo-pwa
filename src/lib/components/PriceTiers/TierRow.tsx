import React, { FunctionComponent } from 'react';

import Price from '@magento/peregrine/lib/Price';

import { TPriceTier } from './remapPriceTiers';

interface ITierRowProps {
    quantity: string;
    item: TPriceTier;
}

const TierRow: FunctionComponent<ITierRowProps> = ({ item, quantity }) => (
    <tr>
        <th scope="row">{quantity}</th>
        <td>
            <Price value={item.value} currencyCode={item.currency} />
        </td>
    </tr>
);

export default TierRow;
