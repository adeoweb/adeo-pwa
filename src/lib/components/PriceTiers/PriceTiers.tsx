import React, { FunctionComponent } from 'react';
import { TPriceRange, TTierPrice } from 'src/lib/types/graphql/Product';
import TierRow from './TierRow';
import remapPriceTiers from 'src/lib/components/PriceTiers/remapPriceTiers';

type TPriceTiersProps = {
    priceTiers?: TTierPrice[];
    priceRange?: TPriceRange;
    isVisible?: boolean;
};

const PriceTiers: FunctionComponent<TPriceTiersProps> = ({
    priceTiers,
    priceRange,
    isVisible = true
}) => {
    if (!isVisible || !priceTiers || !priceTiers.length) {
        return null;
    }

    const items = remapPriceTiers({
        priceTiers,
        priceRange
    });

    if (!items.length) {
        return null;
    }

    return (
        <div className="mt-2">
            <h3>Price tiers</h3>
            <table className="table table-responsive table-hover lh-13">
                <thead>
                    <tr>
                        <th scope="col">Quantity</th>
                        <th scope="col">Price</th>
                    </tr>
                </thead>
                <tbody>
                    {items.map((item, index) => {
                        const quantity =
                            index === items.length - 1
                                ? `${item.quantity}+`
                                : `${item.quantity}-${
                                      items[index + 1].quantity
                                  }`;

                        return (
                            <TierRow
                                key={`tier-${index}`}
                                item={item}
                                quantity={quantity}
                            />
                        );
                    })}
                </tbody>
            </table>
        </div>
    );
};

export default PriceTiers;
