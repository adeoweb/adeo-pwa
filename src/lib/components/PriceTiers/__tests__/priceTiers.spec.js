import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import PriceTiers from '../PriceTiers';
import { IntlProvider } from 'react-intl';

const tier1 = {
    final_price: {
        currency: 'EUR',
        value: 100
    },
    quantity: 2
};

const tier2 = {
    final_price: {
        currency: 'EUR',
        value: 90
    },
    quantity: 4
};

const tier3 = {
    final_price: {
        currency: 'EUR',
        value: 80
    },
    quantity: 6
};

const tier4 = {
    final_price: {
        currency: 'EUR',
        value: 70
    },
    quantity: 8
};

const priceRange = {
    minimum_price: {
        final_price: {
            currency: 'EUR',
            value: 2.5
        }
    }
};

describe('PriceTiers', () => {
    test('renders correctly with one tier', () => {
        const priceTiers = [tier1];

        const tree = createTestInstance(
            <IntlProvider locale="en">
                <PriceTiers priceTiers={priceTiers} priceRange={priceRange} />
            </IntlProvider>
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('renders correctly two tiers', () => {
        const priceTiers = [tier1, tier2];

        const tree = createTestInstance(
            <IntlProvider locale="en">
                <PriceTiers priceTiers={priceTiers} priceRange={priceRange} />
            </IntlProvider>
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('renders correctly with three tiers', () => {
        const priceTiers = [tier1, tier2, tier3];

        const tree = createTestInstance(
            <IntlProvider locale="en">
                <PriceTiers priceTiers={priceTiers} priceRange={priceRange} />
            </IntlProvider>
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('renders correctly with four tiers', () => {
        const priceTiers = [tier1, tier2, tier3, tier4];

        const tree = createTestInstance(
            <IntlProvider locale="en">
                <PriceTiers priceTiers={priceTiers} priceRange={priceRange} />
            </IntlProvider>
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });
});
