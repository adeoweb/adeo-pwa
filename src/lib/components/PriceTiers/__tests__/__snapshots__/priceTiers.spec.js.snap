// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceTiers renders correctly two tiers 1`] = `
<div
  className="mt-2"
>
  <h3>
    Price tiers
  </h3>
  <table
    className="table table-responsive table-hover lh-13"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Quantity
        </th>
        <th
          scope="col"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1-2
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            2
          </span>
          <span>
            .
          </span>
          <span>
            50
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          2-4
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            100
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          4+
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            90
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PriceTiers renders correctly with four tiers 1`] = `
<div
  className="mt-2"
>
  <h3>
    Price tiers
  </h3>
  <table
    className="table table-responsive table-hover lh-13"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Quantity
        </th>
        <th
          scope="col"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1-2
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            2
          </span>
          <span>
            .
          </span>
          <span>
            50
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          2-4
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            100
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          4-6
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            90
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          6-8
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            80
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          8+
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            70
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PriceTiers renders correctly with one tier 1`] = `
<div
  className="mt-2"
>
  <h3>
    Price tiers
  </h3>
  <table
    className="table table-responsive table-hover lh-13"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Quantity
        </th>
        <th
          scope="col"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1-2
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            2
          </span>
          <span>
            .
          </span>
          <span>
            50
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          2+
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            100
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PriceTiers renders correctly with three tiers 1`] = `
<div
  className="mt-2"
>
  <h3>
    Price tiers
  </h3>
  <table
    className="table table-responsive table-hover lh-13"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Quantity
        </th>
        <th
          scope="col"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          1-2
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            2
          </span>
          <span>
            .
          </span>
          <span>
            50
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          2-4
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            100
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          4-6
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            90
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          6+
        </th>
        <td>
          <span>
            €
          </span>
          <span>
            80
          </span>
          <span>
            .
          </span>
          <span>
            00
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
